### PR TITLE
TASK-55927: Add the possibility to delete an ended challenge without announcements

### PR DESCRIPTION
**Pull request by GouadriaHanen**
Add the Delete option for challenge owners to delete ended challenges without announcements

### CODE REVIEW

File: services/src/main/java/org/exoplatform/addons/gamification/storage/RuleStorage.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import org.exoplatform.addons.gamification.service.mapper.EntityMapper;
 import org.exoplatform.addons.gamification.service.mapper.RuleMapper;
 import org.exoplatform.addons.gamification.storage.dao.RuleDAO;
+import org.jgroups.tests.rt.transports.UdpTransport;
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/storage/RuleStorage.java]: Unneeded import, to remove.

File: challenges-webapp/src/main/webapp/vue-app/challenges/challengesServices.js [hunk 2]
```diff
@@ -118,3 +118,14 @@ export function getAllDomains() {
   });
 }
 
+
+export function deleteChallenge(challengeId) {
+  return fetch(`${eXo.env.portal.context}/${eXo.env.portal.rest}/gamification/challenge/api/delete/${challengeId}`, {
+    method: 'PUT',
```
Comment (ahamdi) [inline challenges-webapp/src/main/webapp/vue-app/challenges/challengesServices.js]: Shouldn't we use Delete method ?

File: services/src/main/java/org/exoplatform/addons/gamification/rest/ChallengeRest.java [hunk 3]
```diff
@@ -218,4 +218,38 @@ public Response canAddChallenge() {
...
+  @ApiOperation(value = "check if the current user can add a challenge", httpMethod = "GET", response = Response.class, notes = "This checks if the current user user can add a challenge", consumes = "application/json")
+  @ApiResponses(value = { @ApiResponse(code = 200, message = "challenge deleted"),
+      @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "User not authorized to delete a challenge"),
+      @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+      @ApiResponse(code = HTTPStatus.NOT_FOUND, message = "Object not found"),
+      @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), })
+  public Response deleteChallenge(@ApiParam(value = "challenge id to be deleted", required = true)
+  @PathParam("id")
+  Long challengeId) {
+    if (challengeId ==null || challengeId <= 0 ) {
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/rest/ChallengeRest.java]:
```suggestion
    if (challengeId == null || challengeId <= 0 ) {
```

File: services/src/main/java/org/exoplatform/addons/gamification/storage/RuleStorage.java [hunk 4]
```diff
@@ -107,7 +107,7 @@ public Challenge saveChallenge(Challenge challenge, String username) {
 
   public Challenge getChallengeById(long challengeId) {
     RuleEntity challengeEntity = this.ruleDAO.find(challengeId);
-    if (challengeEntity == null || challengeEntity.getType() == TypeRule.AUTOMATIC) {
+    if (challengeEntity == null || challengeEntity.getType() == TypeRule.AUTOMATIC || challengeEntity.isDeleted()) {
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/storage/RuleStorage.java]: No need for challengeEntity.isDeleted()

File: services/src/main/java/org/exoplatform/addons/gamification/entities/domain/configuration/RuleEntity.java [hunk 5]
```diff
@@ -42,7 +42,7 @@
     @NamedQuery(name = "Rule.getEventList", query = "SELECT rule.event  FROM Rule rule  where rule.type = :type GROUP BY rule.event"),
     @NamedQuery(name = "Rule.deleteRuleByTitle", query = "DELETE FROM Rule rule WHERE LOWER(rule.title) = LOWER(:ruleTitle) "),
     @NamedQuery(name = "Rule.deleteRuleById", query = "DELETE FROM Rule rule WHERE rule.id = :ruleId "),
-    @NamedQuery(name = "Rule.findAllChallengesByUser", query = "SELECT DISTINCT r FROM Rule r where r.audience in (:ids) order by r.endDate desc") })
+    @NamedQuery(name = "Rule.findAllChallengesByUser", query = "SELECT DISTINCT r FROM Rule r where r.isDeleted = false AND r.audience in (:ids) order by r.endDate desc") })
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/entities/domain/configuration/RuleEntity.java]: Is th eproperty isDeleted still needed ?

File: challenges-webapp/src/main/webapp/vue-app/challenges/components/ChallengeCard.vue [hunk 6]
```diff
@@ -33,6 +33,9 @@
                     <v-list-item class="editList" @mousedown="$event.preventDefault()">
                       <v-list-item-title class="editLabel" @click="$emit('edit', challenge)">{{ $t('challenges.edit') }}</v-list-item-title>
                     </v-list-item>
+                    <v-list-item class="editList" @mousedown="$event.preventDefault()">
+                      <v-list-item-title class="editLabel" @click="confirmDelete">{{ $t('challenges.delete') }}</v-list-item-title>
```
Comment (ahamdi) [inline challenges-webapp/src/main/webapp/vue-app/challenges/components/ChallengeCard.vue:40]: Wht is the behaviour when you want to delete a challenge that already have announcements ? does it show an error with an i18n message ?

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: challenges-webapp/src/main/webapp/vue-app/challenges/components/ChallengeCard.vue (hunk 6 above)
Comment (GouadriaHanen) [reply challenges-webapp/src/main/webapp/vue-app/challenges/components/ChallengeCard.vue:40]: yes it displays an error message: "Error deleting challenge : challenge may have announcements or has not ended yet"

**Review by ahamdi — APPROVED** [review]

**Pull request merged**